RFCs/0012-Initialization-of-Adapters-And-Extensions.md

```diff
@@ -0,0 +1,36 @@
+# 0012 Initialization of Test Adapter and Extensions
+
+## Summary
+This note outlines the proposed changes to TestPlatform for handling various scenarios related to
+initialization of adapters and extensions.
+
+## Motivation
+Issues with the current design for initialization of the extensions when multiple versions of same
+extension are involved.
+
+## Problem
+There are 2 main scenarios when there are multiple versions of same adapter competing.
+
+### Upgrade of packages
+User can upgrade the adapter nuget packages in the same VS session.
+In this case, older version of extension had already been initialized and then new extension is sent for
+initialization.
+
+### Multiple versions in the same solution
+User can have projects targeting different version of the test adapter.
+In case of net46 framework where host process is shared, the right adapter does not get initialized.
+
+## Proposed Changes
+Here are a few proposed changes in this regard:
+
+1. All the dependencies including test adapters are copied to the bin directory of the project.
+This is similar to setting /testAdapterPath to bin directory.
+2. Each test container will be hosted in a separate host. An option can be given to the user to choose
```

Comment:
(adding more history for context)

Today VS supports finding adapters from a vsix extension. So even if user doesn't have a adapter nuget package, the adapter from vsix could be used to run it. E.g. [here's](https://marketplace.visualstudio.com/items?itemName=NUnitDevelopers.NUnit3TestAdapter) the one for NUnit framework.

The challenge with this was adapters have a coupling with the test framework version. E.g. mstest v2 will not be able to run mstestv1 references. It was leading to hard to trace errors in such incompatibilities. Secondly, in CI machines, it required VS to be available with the exact same version of adapter installed as vsix. I think few test frameworks have now deprecated vsix adapter extensions.

Nuget packages were easier to maintain (same process as other dev dependencies), provided a way to use the same adapter both in CI and local dev box. But it led to scenarios with multiple adapters being sent for same project.